parlai/core/torch_generator_agent.py

```diff
@@ -1655,15 +1799,34 @@ def __init__(self, k, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.k = k
 
-    def select_paths(self, logprobs, prior_scores, current_length):
+    def select_paths(self, logprobs, prior_scores, current_length) -> _PathSelection:
         values, indices = logprobs.topk(self.k, dim=-1)
         probs = torch.softmax(values, dim=-1)
         choices = torch.multinomial(probs, 1)[:, 0]
         hyp_ids = torch.arange(logprobs.size(0)).to(logprobs.device)
         tok_ids = indices[hyp_ids, choices]
         scores = values[hyp_ids, choices]
         best_scores = prior_scores.expand_as(scores) + scores
-        return (hyp_ids, tok_ids, best_scores)
+
+        vocab_size = logprobs.size()[1]
+        ranks = logprobs.scatter(
+            1,
+            logprobs.argsort(descending=True),
+            torch.arange(vocab_size)
+            .type(torch.FloatTensor)
+            .to(logprobs.device)[None, :]
+            .repeat((self.beam_size, 1))
+            + 1,
+        )
```

Comment:
i think there is actually a bug here -- say beam size is 3 and the tok ids are 1, 2, and 3 -- you are returning the rank of tokens 1,2, and 3 in the first beam. so, instead of `tok_ids.unsqueeze(0)` here you need `tok_ids.unsqueeze(1)`.

see here for example:
```
(Pdb) test  = torch.tensor([[1, 2, 3], [4, 5, 6], [7, 8, 9], [10, 11, 12]])
(Pdb) choice = torch.tensor([1,1,1,1])
(Pdb) test.gather(1, choice.unsqueeze(0))
tensor([[2, 2, 2, 2]])
(Pdb) test.gather(1, choice.unsqueeze(1))
tensor([[ 2],
        [ 5],
        [ 8],
        [11]])
   ```

you want the second thing here